File: 5/5.cpp
#include <iostream>
#include <string>
#include <stack>
#include <cctype>
#include <chrono>

bool annihilates(char &first, char &second)
{
    if (std::tolower(first) == std::tolower(second)) {
        // Base char is the same
        if ((std::isupper(first) && std::islower(second)) ||
            (std::islower(first) && std::isupper(second))) {
            // Polarised
            return true;
        }
    }
    return false;
}

int reactPolymer(std::string &polymer, char ignored = 0)
{
    std::stack<char> stack;
    for (auto &i : polymer)
    {
        if (i == ignored || i == toupper(ignored)){
            continue;
        }
        else if (stack.empty()) {
            stack.push(i);
        }
        else if (annihilates(stack.top(), i)) {
            stack.pop();
        }
        else {
            stack.push(i);
        }
    }
    return stack.size();
}

int main()
{
    std::string polymer;
    std::cin >> polymer;

    // Star 1
    auto start = std::chrono::high_resolution_clock::now();
    int reactedLength = reactPolymer(polymer);

    auto mid = std::chrono::high_resolution_clock::now();
    auto star1Time = std::chrono::duration_cast<std::chrono::duration<double>>(mid - start);
    std::cout << "Star1: Polymer length: " << reactedLength
              << ". Time taken: " << star1Time.count() << std::endl;

    // Star 2
    char minChar = 'a';
    int min = reactedLength;
    
    for (char ch = 'a'; ch <= 'z'; ++ch){
        int length = reactPolymer(polymer, ch);

        //std::cout << new_stack->size() << std::endl;
        if (length < min){
            min = length;
            minChar = ch;
        }
    }
    
    auto finish = std::chrono::high_resolution_clock::now();
    auto star2Time = std::chrono::duration_cast<std::chrono::duration<double>>(finish - mid);
    std::cout << "Star2: Polymer length after removing " << minChar << ": " << min
              << ". Time taken: " << star2Time.count() << std::endl;
}

File: 5/5.py
import sys
import re
from string import ascii_lowercase

def annihilates(first, second):
    # Does this pair of chars annihilate with eachother?
    if first.lower() == second.lower():
        # Same base char
        if (first.isupper() and second.islower()) or \
           (first.islower() and second.isupper()):
           # Polarised
           return True
    return False

def react_polymer(polymer, ignored_chars=''):
    stack = []
    for char in polymer:
        if char in ignored_chars:
            # Star2: We're acting as if these don't exist
            continue
        if len(stack) == 0:
            stack.append(char)
        elif annihilates(stack[-1], char):
            stack.pop()
        else:
            stack.append(char)
    return stack

polymer = sys.stdin.readline()

 # Star 1
stack = react_polymer(polymer)
print("Star1: Polymer of length {} reacts to become length {}".format(len(polymer), len(stack)))

# Star 2
min_ = len(stack)
for char in ascii_lowercase:
    length = len(react_polymer(polymer, ignored_chars=char+char.upper()))
    if length < min_:
        min_ = length
        min_char = char

print("Star2: After removing pair {}, the polymer reacts to become length {}".format(min_char + min_char.upper(), min_))
